{
  Throwable t=threadFailure.getAndSet(null);
  if (t != null) {
    if (t instanceof Error)     throw (Error)t;
 else     if (t instanceof RuntimeException)     throw (RuntimeException)t;
 else     if (t instanceof Exception)     throw (Exception)t;
 else {
      AssertionFailedError afe=new AssertionFailedError(t.toString());
      afe.initCause(t);
      throw afe;
    }
  }
  if (Thread.interrupted())   throw new AssertionFailedError("interrupt status set in main thread");
}
